CMakeLists.txt

```diff
@@ -292,15 +292,19 @@ if(ENABLE_IPV6 AND NOT WIN32)
         CACHE BOOL "Define if you want to enable IPv6 support" FORCE)
   endif()
 
-  if(CMAKE_SYSTEM_NAME STREQUAL "Darwin" AND NOT ENABLE_ARES)
-    set(use_core_foundation_and_core_services ON)
+  if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
+    if (ENABLE_ARES)
+      list(APPEND CURL_LIBS "-licucore")
+    else()
```

Comment:
Isn't `-licucore` necessary for all `APPLE`? The next issue is that this makes it necessary to detect the presence of this OS feature in CMake/autotools, then pass this on via `HAVE_APPLE_IDN` (replacing `__has_include`). Then add this library if the feature is available and enable it in source on this condition too.

Reply:
Sure, we could add it for the other case, too. Or just move it outside the inner condition.
I like the __has_include as it allows to check for a header on compile time.
Feel free to adjust the build scripts.

Reply:
Moving it around will not help for Apple targets not supporting this feature. It will fail when linking a non-existing library.

Reply:
Well, if here is someone who can write CMake conditions and configure checks to check for presents of unicode/uidna.h and whether you can link "icucore". A check for file on disk won't work since the library is in the dylib cache archive. 

On the other side, why bother to check at all. All platforms from Apple support the header file and include the library as far as I know: Apple TV OS, iOS, MacOS, WatchOS and Vision Pro OS. And I verified that the library exists on macOS 10.13, so included for a couple of years already.

Of course it would be nice to have a switch to turn it off in case of someone seeing a problem. But CMake/Configure scripts is outside my expertise.

Reply:
I proposed a patch earlier that added `icucore` unconditionally to CMake. The current logic seems wrong, e.g. what does `ENABLE_ARES` have to do with enabling `icucore`?

But even that simple patch would:
- fail when building with an older (pre v13) Apple toolchain.
- likely fail to build when using a new Apple toolchain but targeting an older OS release.
- doesn't allow to disable this feature. This might have valid reasons like someone hitting a snag with the newly added logic, and still wanting to build curl. Or, someone just doesn't want IDN support for whatever reason.
- leaves this feature broken with autotools. That needs the same detection/enabler/lib-adder logic.

I can't volunteer for this now, because I'm working since Thursday pretty much day and night on open source stuff on my free time.

A way to fix this is to remove header detection from the C code, replace it with `HAVE_APPLE_IDN` (meaning: support is available) and `USE_APPLE_IDN` (we also want to use it) macros, all expected to be set manually for the time being. Then later add the CMake logic doing the automatic detecition/feature-toggle, and the same for autotools.

Reply:
>  what does ENABLE_ARES have to do with enabling icucore?

Well, if you enable ARES, you don't link to CoreFoundation, which links icucore.
So we don't need to reference icucore directly.
But yes, we could change it to use constants as switches.

Reply:
The problem remains that `icucore` has only been available since SDK 13, so this break building curl on everything before. Even with newer SDKs it breaks when explicitly targeting older ones.

There is no other way IMO than making this an option. Then implementing detection for this at the build level.

Reply:
I know, that you'd like to have a switch and I may look into adding one.

How far in past do you need to support older macOS versions? We support 10.13 (from 2017) since that is what current Xcode 15 can build.

Reply:
We don't have this formally defined, I build curl-for-win for 10.9.

IMO this isn't really be about trying to hardwire or guess these versions in the build or source.

This PR already is already good if the feature availability can be passed manually, and offers another option to enable it. We use these mechanics for similar cases.

As a next step (in a separate PR even) detecting availability can be automated. E.g. by setting up the icucore lib and checking for one of the APIs, using `check_function_exists()` in CMake.

Reply:
For curl-for-**win** for **mac**OS 10.9?

We could detect macOS 10.13 SDK (or iOS 11.0 SDK) via #if like this:

```
#if TARGET_OS_OSX

#if MAC_OS_X_VERSION_MIN_REQUIRED < MAC_OS_X_VERSION_10_13
// disable feature for old MacOS SDK
#endif

#elif TARGET_OS_IPHONE

#if __IPHONE_OS_VERSION_MIN_REQUIRED  < __IPHONE_11_0
// disable feature for old iOS SDK (also Watch, VisionOS, AppleTV, etc.)
#endif

#endif
```

So it could be automatically disabled for older SDK.

We'd just need a nice way to detect whether library is there and add it. Configure script needs to get a **-licucore**, too. But not sure where.

Reply:
What's the benefit of scattering detection into these three: a hardwired logic in-source, a dynamic-header detection in-source and future-developed library detection in autotools and CMake?

Reply:
Well, if I could do it all in source, I wouldn't need to wait for someone to do autotools and CMake.
I'd prefer to have it all automatic. If this is opt-in for building on macOS, I fear it may not get used.

Reply:
Pure in-source detection would be nice, but I can't see how this could be technically feasible.

Since we do need detection, our option is to either make it opt-in for now and add detection in a separate commit, or add detection now. We also need an opt-out option in either case.

This is a useful feature, so it's reasonable to assume that even the two-stage implementation will pass through quickly. My worry is autotools, which would need to be done in the same step, but needs a very different skillset than CMake. Worst case we can make an exception in the CI job responsible for keeping CMake and autotools in sync.